library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.datafixerupper.api;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import com.mojang.datafixers.DataFixer;
+import com.mojang.datafixers.DataFixerBuilder;
+import com.mojang.datafixers.schemas.Schema;
+
+import net.minecraft.nbt.NbtCompound;
+
+import org.quiltmc.loader.api.ModContainer;
+import org.quiltmc.qsl.datafixerupper.impl.QuiltDataFixesInternals;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Provides methods to register custom {@link DataFixer}s.
+ */
+public final class QuiltDataFixes {
+	private QuiltDataFixes() {
+		throw new RuntimeException("QuiltDataFixes only contains static declarations.");
+	}
+
+	/**
+	 * A "base" version {@code 0} schema, for use by all mods.
+	 * <p>
+	 * This schema <em>must</em> be the first one added!
+	 *
+	 * @see DataFixerBuilder#addSchema(int, BiFunction)
+	 */
+	public static final BiFunction<Integer, Schema, Schema> BASE_SCHEMA = (version, parent) -> {
+		checkArgument(version == 0, "version must be 0");
+		checkArgument(parent == null, "parent must be null");
+		return QuiltDataFixesInternals.createBaseSchema();
+	};
+
+	/**
+	 * Registers a new data fixer.
+	 *
+	 * @param modId          the mod identifier
+	 * @param currentVersion the current version of the mod's data
+	 * @param dataFixer      the data fixer
+	 */
+	public static void registerFixer(String modId, int currentVersion, DataFixer dataFixer) {
+		checkNotNull(modId, "modId cannot be null");
+		checkArgument(currentVersion >= 0, "currentVersion must be positive");
+		checkNotNull(dataFixer, "dataFixer cannot be null");
+
+		if (isFrozen()) {
+			throw new IllegalStateException("Can't register data fixer after registry is frozen");
+		}
+
+		QuiltDataFixesInternals.registerFixer(modId, currentVersion, dataFixer);
+	}
+
+	/**
+	 * Registers a new data fixer.
+	 *
+	 * @param mod            the mod container
+	 * @param currentVersion the current version of the mod's data
+	 * @param dataFixer      the data fixer
+	 */
+	public static void registerFixer(ModContainer mod, int currentVersion, DataFixer dataFixer) {
```

Comment:
Ok, sorry for not mentioning it earlier, but new day new thoughts

What if the current version was something declared in the QMJ so it would be easy to locate?
Would keep this overload still, but it's an interesting thought to have.

Reply:
So something like `"data_version": int`? Hm...

Reply:
Er, wait, problem: the `DataFixerBuilder` constructor takes the latest data version... meaning if a user declares it in their `quilt.mod.json`, they'd need to read it themselves anyway.

Reply:
This method has the ModContainer tho?

Reply:
```java
public static void initialize(ModContainer mod) {
    QuiltDataFixerBuilder builder = new QuiltDataFixerBuilder(DATA_VERSION);
    // END-USER code that needs the data version
    builder.addSchema(0, QuiltDataFixes.BASE_SCHEMA);
    // etc, etc...
    
    QuiltDataFixes.registerFixer(mod, DATA_VERSION, builder.build(Util::getBootstrapExecutor));
    // DEV code that needs the data version
}
```

Reply:
ooooh

Now we see.